job-service-dropwizard/src/main/java/com/hpe/caf/services/job/dropwizard/health/PortsHealthCheck.java

```diff
@@ -0,0 +1,111 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hpe.caf.services.job.dropwizard.health;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.jetty.server.ServerConnector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.codahale.metrics.health.HealthCheck;
+
+import io.dropwizard.core.setup.Environment;
+
+public final class PortsHealthCheck extends HealthCheck
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(PortsHealthCheck.class);
+
+    private final Environment environment;
+    private final Type type;
+
+    public enum Type
+    {
+        ALIVE,
+        READY;
+    }
+
+    public PortsHealthCheck(final Environment environment, final Type type)
+    {
+        this.environment = environment;
+        this.type = type;
+    }
+
+    @Override
+    protected Result check()
+    {
+        final List<ServerConnector> serverConnectors = Arrays
+                .stream(environment.getApplicationContext().getServer().getConnectors())
+                .filter(connector -> connector instanceof ServerConnector)
+                .map(connector -> (ServerConnector)connector)
+                .toList();
+
+        if (serverConnectors.isEmpty()) {
+            final String message = "No ServerConnectors found";
+            LOGGER.error(message);
+            return Result.unhealthy(message);
+        }
+
+        for (final ServerConnector serverConnector : serverConnectors) {
+            final String serverConnectorName = serverConnector.getName();
+            final int serverConnectorPort = serverConnector.getPort();
+
+            switch (type) {
+                case ALIVE:
+                    if (!serverConnector.isStarted()) {
+                        final String message = getMessage(serverConnectorName, serverConnectorPort, "not started");
+                        LOGGER.error(message);
+                        return Result.unhealthy(message);
+                    }
+
+                    if (!serverConnector.isOpen()) {
+                        final String message = getMessage(serverConnectorName, serverConnectorPort, "not open");
+                        LOGGER.error(message);
+                        return Result.unhealthy(message);
+                    }
+
+                    LOGGER.error(getMessage(serverConnectorName, serverConnectorPort, "started and open")); // TODO debug
+
+                    break;
+                case READY:
+                    if (!serverConnector.isAccepting()) {
```

Comment:
I though putting the `!serverConnector.isAccepting()` check in the readiness checks would prevent restarting the service in the case where the service was just busy with other requests. 

Looking at the issue described in https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=914157, I am not sure what state the connectors would have been in when this issue happened.

If they had been not started or closed, then these changes would mean the service would have been restarted, which is what we want.

If they have been open but not accepting, then putting this in the readiness check and not in the liveness check means we would not be restarting the service if this issue happened again. However, putting the ``!serverConnector.isAccepting()` check in the liveness check means that we would restart the service if this issue happened again, but also that we may potentially be restarting the service when it is just busy with other requests (killing those requests), so I am not sure what the best thing to do here is?

Reply:
I'm not sure, but I suspect it was closed, so I think you're changes cover it, hopefully.